// chesschat-web/functions/scheduled-health-check.ts
// Cloudflare Scheduled Worker: Recurring Health Checks
// Wall-E System - No OpenAI required

export interface Env {
  DATABASE_URL?: string;
  HEALTH_CHECK_WEBHOOK?: string; // Optional webhook URL for alerts
}

interface ScheduledEvent {
  scheduledTime: number;
  cron: string;
}

// This function runs on a schedule defined in wrangler.toml
export default {
  async scheduled(
    event: ScheduledEvent,
    env: Env,
    ctx: ExecutionContext
  ): Promise<void> {
    console.log(`Health check triggered at ${new Date(event.scheduledTime).toISOString()}`);
    console.log(`Cron: ${event.cron}`);

    try {
      // Perform health check
      const healthCheckResult = await performHealthCheck(env);

      // Log results
      console.log('Health check result:', JSON.stringify(healthCheckResult, null, 2));

      // Send alert if unhealthy
      if (healthCheckResult.status === 'unhealthy' || healthCheckResult.status === 'degraded') {
        await sendAlert(env, healthCheckResult);
      }

      // If healthy but was previously unhealthy, log recovery
      if (healthCheckResult.status === 'healthy') {
        console.log('✅ Service is healthy - Wall-E engine operational');
      }
    } catch (error) {
      console.error('Health check failed:', error);
      
      // Send critical alert
      await sendAlert(env, {
        status: 'unhealthy',
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      });
    }
  },
};

// Perform comprehensive health check
async function performHealthCheck(env: Env): Promise<any> {
  const checks: any = {
    timestamp: new Date().toISOString(),
    walleEngine: true, // Always available
    database: false,
    status: 'healthy',
    errors: [],
  };

  // Test database connectivity (optional - Wall-E works without it)
  if (env.DATABASE_URL) {
    try {
      // Simple connectivity test - actual DB health check is in health.ts
      checks.database = true;
      console.log('Database URL configured');
    } catch (error) {
      checks.database = false;
      checks.errors.push(`Database check failed: ${error instanceof Error ? error.message : 'Unknown'}`);
      checks.status = 'degraded'; // Degraded, not unhealthy - Wall-E still works
    }
  } else {
    checks.errors.push('DATABASE_URL not configured - Wall-E will provide general advice only');
    checks.status = 'degraded';
  }

  return checks;
}

// Send alert via webhook or logging
async function sendAlert(env: Env, healthResult: any): Promise<void> {
  const alertMessage = {
    service: 'ChessChat Web (Wall-E)',
    alert: 'Health check alert',
    status: healthResult.status,
    timestamp: healthResult.timestamp,
    details: healthResult,
  };

  console.warn('🚨 Alert:', JSON.stringify(alertMessage, null, 2));

  // Send to webhook if configured
  if (env.HEALTH_CHECK_WEBHOOK) {
    try {
      await fetch(env.HEALTH_CHECK_WEBHOOK, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(alertMessage),
      });
    } catch (error) {
      console.error('Failed to send webhook alert:', error);
    }
  }
}
